bundle/Routing/SiteaccessResolver.php

```diff
@@ -0,0 +1,388 @@
+{
+    private $persistenceHandler;
+    private $excludedSiteaccessSet;
+    private $excludedSiteaccessGroupSet;
+
+    /**
+     * @var ConfigResolverInterface
+     */
+    private $configResolver;
+
+    /**
+     * @var SiteAccess
+     */
+    private $currentSiteaccess;
+
+    /**
+     * @var array
+     */
+    private $siteaccesses;
+
+    /**
+     * @var array
+     */
+    private $siteaccessGroupsBySiteaccess;
+
+    /**
+     * @var array
+     */
+    private $siteaccessRootLocationIdMap;
+
+    /**
+     * @var array
+     */
+    private $locationIdSiteaccessesMap = [];
+
+    public function __construct(
+        Handler $persistenceHandler,
+        array $excludedSiteaccesses,
+        array $excludedSiteaccessGroups
+    ) {
+        $this->persistenceHandler = $persistenceHandler;
+        $this->excludedSiteaccessSet = array_fill_keys($excludedSiteaccesses, true);
+        $this->excludedSiteaccessGroupSet = array_fill_keys($excludedSiteaccessGroups, true);
+    }
+
+    /**
+     * @param ConfigResolverInterface $configResolver
+     */
+    public function setConfigResolver(ConfigResolverInterface $configResolver): void
+    {
+        $this->configResolver = $configResolver;
+    }
+
+    public function setSiteaccess(SiteAccess $currentSiteAccess = null): void
+    {
+        $this->currentSiteaccess = $currentSiteAccess;
+    }
+
+    public function setSiteaccessList(array $siteaccesses): void
+    {
+        $this->siteaccesses = $siteaccesses;
+    }
+
+    public function setSiteaccessGroupsBySiteaccess(array $siteaccessGroupsBySiteaccess): void
+    {
+        $this->siteaccessGroupsBySiteaccess = $siteaccessGroupsBySiteaccess;
+    }
+
+    public function resolve(Location $location): string
+    {
+        $siteaccesses = $this->getSiteaccesses($location);
+
+        if (empty($siteaccesses)) {
+            return $this->currentSiteaccess->name;
+        }
+
+        try {
+            $availableLanguageSet = $this->getLanguageSet($location);
+        } catch (Exception $e) {
+            return $this->currentSiteaccess->name;
+        }
+
+        $currentSiteaccess = $this->currentSiteaccess->name;
+        $alwaysAvailable = $location->contentInfo->alwaysAvailable;
+        $currentPrioritizedLanguages = $this->getCurrentPrioritizedLanguages();
+
+        $maybeSiteaccess = $this->matchTranslationSiteaccess(
+            in_array($currentSiteaccess, $siteaccesses, true) ? $currentSiteaccess : null,
+            $currentPrioritizedLanguages,
+            $availableLanguageSet,
+            $alwaysAvailable
+        );
+
+        if ($maybeSiteaccess !== null) {
+            return $maybeSiteaccess;
+        }
+
+        $currentPrimaryLanguage = reset($currentPrioritizedLanguages);
+        $siteaccessesWithSamePrimaryLanguage = $this->matchSiteaccessesByPrimaryLanguage($siteaccesses, $currentPrimaryLanguage);
+
+        foreach ($siteaccessesWithSamePrimaryLanguage as $siteaccess) {
```

Comment:
It should be possible to merge this with the block below, as it's basically a special case of it.